yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,981 @@
+        flushCurrentEventsToATS
+        false
+    }
+  }
+
+  /**
+   * If the event reaches the batch size or flush is true, push events to ATS.
+   *
+   * @param event event. If null no event is queued, but the post-queue flush logic
+   *              still applies
+   * @return true if the event should stop the service.
+   */
+  private def handleEvent(event: HandleSparkEvent): Unit = {
+    var push = false
+    // if we receive a new appStart event, we always push
+    // not much contention here, only happens when service is stopped
+    entityLock synchronized {
+      logDebug(s"Processing event $event")
+      if (eventsProcessed.get() % 1000 == 0) {
+        logDebug(s"${eventsProcessed} events are processed")
+      }
+      val sparkEvent = event.sparkEvent
+      sparkEvent match {
+        case start: SparkListenerApplicationStart =>
+          // we already have all information,
+          // flush it for old one to switch to new one
+          logDebug(s"Handling application start event: $event")
+          appName = start.appName
+          userName = start.sparkUser
+          startTime = start.time
+          if (startTime == 0) {
+            startTime = now()
+          }
+          setContextAppAndAttemptInfo(
+            start.appId,
+            start.appAttemptId)
+          // flush old entity
+          resetCurrentEntity
+          appStartEventProcessed.set(true)
+          appEndEventProcessed.set(false)
+          // force create the new one
+          val en = started(getOrCreateCurrentEntity, startTime)
+          logInfo(s"Application started: $event")
+          push = true
+
+
+        case end: SparkListenerApplicationEnd =>
+          if (!appEndEventProcessed.get()) {
+            // we already have all information,
+            // flush it for old one to switch to new one
+            logInfo(s"Application end event: $event")
+            if (!appStartEventProcessed.get()) {
+              logWarning(s"Handling application end event without application start $event")
+            }
+            // flush old entity
+            val endtime = if (end.time > 0) end.time else now()
+            if (startTime == 0) {
+              // special case: an app end is received without any active start
+              // event sent yet
+              logInfo(s"start time is 0; fixing up to match end time")
+              startTime = endtime
+            }
+            // reset the current entity
+            resetCurrentEntity
+            // then trigger its recreation
+            val en = completed(getOrCreateCurrentEntity(),
+              startTime,
+              endtime,
+              sparkApplicationId,
+              sparkApplicationAttemptId)
+            appStartEventProcessed.set(false)
+            appEndEventProcessed.set(true)
+            push = true
+          }
+        case _ =>
+          val currentEntity: TimelineEntity = getOrCreateCurrentEntity()
```

Comment:
Also you're calling `getOrCreateCurrentEntity` in L732 again, so is this even necessary?